### PR TITLE
feat: Create Playlist feature

### PR DESCRIPTION
**Pull request by Miralthor**
# Create Playlist Feature

Adds a "Create Playlist" feature that allows creating playlists from various sources — not just syncing existing ones.

## Features

- **File Import** with drag & drop (CSV, TXT, TSV, M3U)
- **Text Import** (copy-paste: `Artist - Title`)
- **URL Import** (Spotify, Apple Music, YouTube Music, etc.)
- **Auto-matching** with confidence scores and candidate picker for review
- **Append mode** to extend existing playlists (no duplicates)
- **Configurable default source** in settings
- **Imports list** with resume/delete functionality

## Technical

- 61+ tests
- 15 languages localized
- Crash recovery for interrupted imports
- Server-side limits (10MB upload, 10k tracks)

### CODE REVIEW

File: songmirror/engine/targets/amazon_music.py [hunk 1]
```diff
@@ -709,6 +709,46 @@ def resolve(self, track, cache):
...
         return best, "search"
 
+    def search_candidates(self, query, *, limit=5):
+        query = str(query or "").strip()
+        if not query:
+            return []
+        try:
+            rows = self._search("name", query, limit=max(limit, 1))
+        except Exception:
+            return []
```
Comment (ahnafnafee) [inline songmirror/engine/targets/amazon_music.py:723]: **[P2] Propagate authentication and transport failures instead of caching misses**

`resolve()` now calls `search_candidates()`, whose broad catch converts `TargetAuthError` and transport failures into an empty result. `resolve()` then persists `None` in the search cache, and later attempts return that cached miss without contacting the provider. In a reproduction where the first lookup raised `TargetAuthError` and the next response would succeed, two resolutions produced only one provider lookup. This also violates the fatal-error contract documented on `TargetAuthError`.

Recommended fix: let operational errors propagate through this wrapper, or explicitly distinguish genuine empty search results from authentication/transient/transport failures. Only cache a miss after a successful catalog request. Add a test that raises an authentication or transient error, verifies the cache has no new miss, then succeeds after recovery. Audit the equivalent broad catches in the new Apple search wrappers for the same propagation issue.

File: frontend/src/pages/Imports.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                      <ServiceLogo service={logoId} className={`size-5 ${tagText(logoId)}`} />
+                    ) : (
+                      <LuHistory className="size-5 text-text-3" aria-hidden="true" />
+                    )}
+                  </div>
+                  <div className="min-w-0">
+                    <div className="truncate font-medium text-text">{job.destination_name}</div>
+                    <div className="mt-0.5 text-sm text-text-3">
+                      {sourceKindLabel(job.source_kind)} • {formatNumber(job.total_tracks)} {t('tracks')} •{' '}
+                      {new Date(job.created_at).toLocaleDateString()}
```
Comment (ahnafnafee) [inline frontend/src/pages/Imports.tsx]: **[P3] Use the shared plural and date formatters**

Concatenating the number with `t('tracks')` renders “1 tracks” and prevents languages such as Arabic from choosing their required plural categories. `toLocaleDateString()` follows the browser locale rather than the language selected inside the application. Both bypass the formatting rules in `frontend/README.md`.

Recommended fix: use `formatTrackCount(job.total_tracks)` and `formatDateTime(job.created_at)` from `src/lib/format.ts` (with explicit date options if needed). Check singular/plural rendering and switch the application language while the browser locale stays unchanged.

File: frontend/src/pages/CreatePlaylist.tsx [hunk 3]
```diff
@@ -0,0 +1,1337 @@
...
+      void (async () => {
+        try {
+          const data = await importApi.getImport(jobId)
+          if (cancelled) return
+          setJob(data.job)
+          setJobData(data)
+
+          if (data.job.status === 'ready') {
+            const full = await fetchImportWithAllTracks(jobId)
+            if (cancelled) return
```
Comment (ahnafnafee) [inline frontend/src/pages/CreatePlaylist.tsx]: **[P1] Finish loading review data before publishing the ready status**

When polling returns `ready`, `setJob(data.job)` changes the effect's `jobStatus` dependency. React runs the cleanup, setting `cancelled = true`, while `fetchImportWithAllTracks()` is pending. The guard then discards the full response before `setStep('review')`, and the next effect does not poll a ready job. Successful imports remain on the progress screen instead of exposing the promised candidate review. Confirmed against the production bundle with delayed mocked API responses.

Recommended fix: handle the `ready` branch by fetching all tracks first, checking cancellation, and then committing `job`, `jobData`, and `step` together. Alternatively, derive the wizard step from job status while keeping the full-data load independent of the polling effect. Add a browser regression covering `matching -> ready` with nonzero response latency and assert that Review Matches appears.

File: songmirror/services/import_parsers/file.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        # Prefer CSV when the file clearly looks columnar; otherwise treat as text.
+        sample = text.lstrip()[:2048]
+        looks_columnar = (
+            ("\n" in sample or "\r" in sample)
+            and any(delimiter in sample for delimiter in (",", ";", "\t"))
+            and sample.count(",") + sample.count(";") + sample.count("\t") >= 2
+        )
+        if looks_columnar:
+            try:
+                result = parse_csv(text)
```
Comment (ahnafnafee) [inline songmirror/services/import_parsers/file.py:66]: **[P2] Do not classify ordinary TXT entries as CSV solely because they contain commas**

Two commas anywhere in a multiline TXT file satisfy this heuristic. For example:

```text
Earth, Wind & Fire - September
Earth, Wind & Fire - Fantasy
```

The file parser produces artist `Earth` and titles `Wind & Fire - September` / `Wind & Fire - Fantasy`, without warnings. Pasting the identical content correctly preserves the artist and titles. This breaks the advertised TXT import format and feeds incorrect metadata into matching.

Recommended fix: honor TXT as plain text by default, or require stronger evidence such as recognized column headers before switching it to CSV. Comma frequency alone is insufficient. Add this file-versus-paste equivalence case and keep coverage for genuine CSV/TSV files.

File: songmirror/services/imports.py [hunk 5]
```diff
@@ -0,0 +1,1607 @@
...
+                    )
+                )
+                already_present += 1
+                continue
+
+            pending_ids.append(target_id)
+            pending_positions.append(int(track["position"]))
+
+            # Flush in modest batches so a pause can stop between chunks.
+            if len(pending_ids) >= 50:
```
Comment (ahnafnafee) [inline songmirror/services/imports.py]: **[P2] Deduplicate IDs that are already queued in the current batch**

The earlier duplicate check only consults `existing_ids`, which is updated after a batch is written. Two input rows resolving to the same ID within one batch therefore both enter `pending_ids`. A reproduction appending two rows resolved to `t1` wrote `t1` twice and reported two additions and zero skips. Providers such as Spotify allow those duplicate writes, so this violates the promised duplicate-free append behavior.

Recommended fix: also track IDs queued in the current batch and associate repeated input positions with one provider write. Mark duplicate positions as already present only after that write succeeds, so a failed write remains retryable. Cover duplicates within one batch, across the 50-item boundary, and IDs already present in the destination.

File: songmirror/services/import_matching.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+        usable = [candidate for candidate in (candidates or []) if candidate and _candidate_id(candidate)]
+        if not usable:
+            return None
+
+        scored = self._score_candidates(source, usable)
+        if scored:
+            best = scored[0]
+            best.reason = "isrc_match"
+            best.score = max(best.score, 0.99)
+            best.acceptable = True
```
Comment (ahnafnafee) [inline songmirror/services/import_matching.py]: **[P2] Apply recording compatibility checks before accepting identity shortcuts**

This branch forces `acceptable = True` and at least 99% confidence even when the scoring helper rejected the recording. In a reproduction, an ISRC candidate named `A Song (Live)` with a conflicting duration received `(0.0, False)` from `score_candidate`, yet was selected as an exact 99% match without searching the correct catalog candidate. The cache-hit branch similarly accepts conflicting-duration recordings at 95% confidence. These shortcuts bypass the existing recording safeguards and the user's review step.

Recommended fix: apply `recording_metadata_compatible` (or the existing `compatible_isrc_candidates` helper for ISRC results) before automatically accepting a candidate. On a known performer/version/duration conflict, fall through to catalog search or require review instead of boosting confidence. Treat explicitly manual mappings separately from automatic cache entries. Add tests for cached and ISRC-backed live/studio and duration conflicts.

File: songmirror/services/settings.py [hunk 7]
```diff
@@ -71,10 +83,16 @@ def _read(self):
             return {}
 
     def load(self):
-        return dict(self._data)
+        data = dict(DEFAULT_SETTINGS)
+        data.update(self._data)
+        return data
```
Comment (ahnafnafee) [inline songmirror/services/settings.py:88]: **[P2] Preserve persisted-state detection for legacy credential migration**

`load()` now returns a nonempty defaults dictionary even when the store has no persisted settings. `AccountProfileStore._copy_legacy_settings()` uses `if not local.load()` to decide whether to seed a profile, so it no longer copies legacy credentials into the profile's settings file. The existing migration test fails on this commit and passes on the base. The root fallback still supplies credentials at runtime, but the intended migration copy is skipped.

Recommended fix: separate effective defaults from persisted settings. For example, preserve `load()` as the persisted mapping and apply UI defaults at the settings API/getter boundary, or add an explicit persisted-state query and use it in migration. Keep the existing test verifying that an empty profile receives its settings file, and verify that an already populated profile is not overwritten.

**Review by ahnafnafee — CHANGES_REQUESTED** [review]
Requesting changes on commit `d6d8370ada33ff38b5a08cf5e6e837e53950aa07`. Successful imports can get stuck before the review step, and there are correctness issues in duplicate handling, recording selection, TXT parsing, provider error propagation, and credential migration. The seven inline comments describe the triggers, observed behavior, and recommended fixes; the localization comment is lower priority.

Validation performed on this commit:

- `python -m pytest -q`: **976 passed, 2 failed, 1 skipped**. The failures are `test_legacy_provider_settings_migrate_to_a_stable_default_profile` and `test_resolvers_refresh_old_automatic_matches_and_require_the_correct_recording[None-amazon]`. Both cases pass on the base commit, `efcfe13ad22e0f54cf9dee509726e78984001d73`.
- The credential-migration failure is explained by the inline settings comment. The Amazon test failure itself is an outdated test double: `lambda *_` does not accept the newly supplied `limit=` keyword. Update the fake to accept the real signature, and add a separate test proving provider failures propagate without creating a cached miss.
- Frontend build, lint, and the existing browser suite passed. A focused production-bundle Playwright reproduction with delayed mocked import responses nevertheless confirmed that matching completion does not open the review step.
- `pnpm -C frontend i18n:check` fails with `Obsolete English message Approve All: run pnpm i18n:extract`; the same check passes on the base. Regenerate the catalogs, reconcile obsolete/new entries across locales, and rerun the check.

The functional reproductions used mocked providers and temporary data. Please add focused regression coverage for the affected paths described inline and rerun the existing checks after fixing them.

Optional cleanup: the five new convenience methods in `songmirror/engine/targets/base.py` (`create_playlist`, `get_playlist_info`, `get_playlist_tracks`, `add_track_to_playlist`, and `track_exists_in_playlist`) have no production callers. This is a possible case of speculative generality: remove them unless the import workflow will actually use them, so the provider contract does not acquire a second unused interface.

**Comment by Miralthor** [comment]
Hi, thanks for the thorough review. Sorry for the messy git history, I had some local branch issues to sort out and the force-pushes got out of hand.

All seven findings have been addressed:

**1. [P1] Matching→ready React race condition** — Restructured the polling effect so that on `ready` status, the full track fetch completes *before* publishing job state and step. No premature `setJob()` triggers the cleanup race anymore.

**2. [P2] Provider error propagation** — Updated `search_candidates()` and `search_by_isrc()` in both `amazon_music.py` and `apple.py` to re-raise `TargetAuthError` and `TargetTransientError` instead of catching them into empty results. Only genuine empty catalog responses are returned as `[]`.

**3. [P2] Shared formatters in Imports page** — Replaced raw `toLocaleDateString()` and manual track-count concatenation with `formatTrackCount()` and `formatDateTime()` from `@/lib/format`. Regenerated locale catalogs and removed the obsolete `Approve All` key.

**4. [P2] TXT file misclassified as CSV** — `.txt`/`.text` files now always use `parse_text()`, matching paste behavior exactly. CSV sniffing is retained only for extensionless uploads.

**5. [P2] Batch-internal duplicate target IDs** — Refactored pending writes from parallel lists to `pending_id_to_positions: dict[str, list[int]]`. Provider receives each ID once; duplicate input positions share the write result. First position gets `added`, later duplicates get `already_present` on success, or all stay `failed` on provider error.

**6. [P2] Recording compatibility for ISRC/cache shortcuts** — ISRC and cache-hit paths now apply `recording_metadata_compatible` before auto-accepting. Conflicting duration (>2s) or artist metadata falls through to ambiguous/review status instead of being boosted to exact match.

**7. [P2] Legacy credential migration** — Added `has_persisted_values()` to `SettingsStore` (returns `bool(self._data)`). `_copy_legacy_settings()` now uses this instead of truthiness of `load()`, so empty-profile migration works correctly while `load()` still returns effective defaults.

**Optional cleanup** — Removed the 5 unused convenience methods (`create_playlist`, `get_playlist_info`, `get_playlist_tracks`, `add_track_to_playlist`, `track_exists_in_playlist`) from `engine/targets/[base.py](https://github.com/Users/niktruninger/dev/songmirror/engine/targets/base.py)` and their dedicated test.

**Validation:**

- `python -m pytest -q`: **988 passed** (was 976 passed, 2 failed)
- `pnpm -C frontend build`: ✅
- `pnpm -C frontend i18n:check`: ✅ (15 languages, 1137 English messages)

Happy to address any follow-ups.

File: .claude-flow/policy/state.json
Comment (ahnafnafee) [inline .claude-flow/policy/state.json:1]: Remove these user-centric files from the PR

File: .claude-flow/neural/stats.json
Comment (ahnafnafee) [inline .claude-flow/neural/stats.json:1]: Remove these user-centric files from the PR

File: .tools/env
Comment (ahnafnafee) [inline .tools/env:1]: Remove these user-centric files from the PR

File: .tools/env.fish
Comment (ahnafnafee) [inline .tools/env.fish:1]: Remove these user-centric files from the PR

File: .tools/uv
Comment (ahnafnafee) [inline .tools/uv:1]: Remove these user-centric files from the PR

File: .tools/uvx
Comment (ahnafnafee) [inline .tools/uvx:1]: Remove these user-centric files from the PR

File: songmirror/engine/targets/amazon_music.py [hunk 8]
```diff
@@ -709,6 +709,54 @@ def resolve(self, track, cache):
...
+        if not query:
+            return []
+        try:
+            rows = self._search("name", query, max(limit, 1))
+        except TargetAuthError:
+            raise
+        except TargetTransientError:
+            raise
+        except Exception:
+            return []
```
Comment (ahnafnafee) [inline songmirror/engine/targets/amazon_music.py:723]: **[P2] Propagate the actual provider transport failures**

The fix now re-raises `TargetAuthError` and `TargetTransientError`, but Amazon's actual request layer still raises ordinary `requests.ConnectionError` and `HTTPError`. These fall into the remaining broad catch, become empty search results, and are persisted as permanent misses by `resolve()`.

Reproduced through the real `_request -> _search -> resolve` path with only HTTP transport mocked: both a connection failure and an HTTP 503 produced `(None, "search")`, stored a cache miss, and prevented a second resolution from making another request after recovery. Apple has the analogous problem after exhausted transport retries. The Qobuz and TIDAL search wrappers also still swallow `TargetAuthError`.

Recommended fix: normalize operational failures to the appropriate target exception at the request boundary, or propagate the actual request exceptions through these wrappers. A failed catalog request must remain distinct from a successful request with no matches. Apply the propagation rule consistently across providers. Add regression tests through the real request path for a connection error/503 followed by recovery, asserting that no negative cache entry is created and the next attempt contacts the provider.

File: songmirror/services/import_matching.py [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+        cached_id = (self.cache.get("search") or {}).get(cache_key)
+        if cached_id:
+            validated = self._validate_target_id(str(cached_id), source)
+            if validated is not None:
+                if recording_metadata_compatible(source, _compat_view(validated)):
+                    candidate = self._to_candidate(
+                        validated,
+                        score=0.95,
+                        reason="cache_hit",
+                        acceptable=True,
```
Comment (ahnafnafee) [inline songmirror/services/import_matching.py:231]: **[P2] Validate automatic cache entries against destination metadata**

The compatibility check still does not validate production cache hits. All seven production target classes inherit `MirrorTarget.fetch_track()`, which returns `None`. `_validate_target_id()` then fills the candidate's title, artist, and duration from the source track, so this check compares the source against a copy of itself and passes.

Using the real Spotify adapter with an automatic cache entry pointing at a live recording, the result was still `exact` at 95% confidence, with zero catalog queries. The displayed metadata was copied from the requested studio recording. The new tests pass because their fake implements a metadata lookup that none of the production adapters provides. The ISRC compatibility fix does work; this remaining gap is specifically the automatic cache path.

Recommended fix: implement destination metadata lookup for cached IDs, or treat an unvalidated automatic cache entry as a search hint and fall through to catalog search/review. Do not manufacture compatibility evidence from source metadata. Keep explicitly manual mappings distinct from automatic cache entries. Add coverage using a production adapter or a fake that inherits the default `fetch_track()`, with a conflicting cached live/studio recording and unavailable metadata.

**Review by ahnafnafee — CHANGES_REQUESTED** [review]
Follow-up review of `7cefa01c711be228d23a3b6f26f8ae951489ee58`: requesting changes because two original correctness findings remain partially fixed, and the update introduces unrelated local artifacts.

Five original findings are confirmed fixed: the matching-to-review polling race, TXT comma parsing, duplicate writes within a batch, legacy credential migration, and Imports history localization. The ISRC compatibility check now works, and the five unused target wrappers were removed. The two code comments explain the remaining provider-error and automatic-cache validation gaps, including reproductions and recommended fixes.

Please also remove `.tools/` and `.claude-flow/` from the PR. Together they add approximately **51.4 MB** of local executables and generated state, including a **402,132-line** policy-state file. These are unrelated to the application feature. Keep them local and add appropriate ignore rules to prevent them from being committed again. The existing six file-cleanup comments are retained in this review.

Validation on this head:

- Backend: **987 passed, 1 skipped**.
- Frontend build, lint, translation/catalog checks, and the existing browser suite passed.
- The original delayed-response production-bundle reproduction now reaches Review Matches.
- Additional browser checks confirmed selected-language date/count/status formatting and resuming/paginating a 205-track import.
- Provider and cache reproductions used mocked transport or catalog responses with the production adapter paths; no live playlists were modified.

Please add focused regression coverage for the remaining production-path gaps and rerun these checks after addressing them.

**Review by ahnafnafee — CHANGES_REQUESTED** [review]
Please cleanup these issues

**Comment by Miralthor** [comment]
Hi, all findings have been addressed:

- **Provider error propagation**: Transport failures (ConnectionError, 429, 5xx, etc.) are now normalized to `TargetTransientError` at each adapter's request boundary across all providers (Amazon web+API, Apple, Qobuz, Tidal, Deezer, Spotify, YTMusic). Only genuine empty catalog responses are cached as misses.
- **Cache validation**: Automatic cache hits without independently fetched destination metadata fall through to catalog search/review instead of being auto-accepted. Manual user mappings are preserved as exact selections. `MATCHING_CACHE_VERSION` bumped to purge stale automatic entries.
- **Apple mutation safety**: Transient failures in `_add_one` now verify write landed before retry (except 429, which proves mutation was not accepted).
- **React polling race**: Full track fetch completes before publishing `ready` state.
- **Batch duplicates**: Provider receives each target ID once; later duplicates share the write result.
- **Recording compatibility**: ISRC/cache shortcuts apply `recording_metadata_compatible` before auto-accepting.
- **Legacy credential migration**: Added `has_persisted_values()` to distinguish persisted settings from effective defaults.
- **TXT parsing**: `.txt`/`.text` files always use `parse_text()`, matching paste behavior.
- **Formatters**: Imports page uses `formatTrackCount()` and `formatDateTime()`.
- **Cleanup**: Removed `.tools/` and `.claude-flow/` from git, added ignore rules.

**Review by ahnafnafee — CHANGES_REQUESTED** [review]
Edit: Resolved

**Review by ahnafnafee — APPROVED** [review]
LGTM
